lib/events/encryptionevent.h

```diff
@@ -0,0 +1,101 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#pragma once
+
+#include "stateevent.h"
+#include "eventcontent.h"
+
+namespace QMatrixClient
+{
+    class EncryptionEventContent: public EventContent::Base
+    {
+        public:
+            enum EncryptionType : size_t { MegolmV1AesSha2 = 0,
+                                           Undefined };
+
+            explicit EncryptionEventContent(EncryptionType et = Undefined)
+                : encryption(et)
+            { }
+            explicit EncryptionEventContent(const QJsonObject& json);
+
+            EncryptionType encryption;
+            QString algorithm;
+            int rotationPeriodMs;
+            int rotationPeriodMsgs;
+
+        protected:
+            void fillJson(QJsonObject* o) const override;
+    };
+
+    using EncryptionType = EncryptionEventContent::EncryptionType;
+
+    class EncryptionEvent : public StateEvent<EncryptionEventContent>
+    {
+            Q_GADGET
+        public:
+            DEFINE_EVENT_TYPEID("m.room.encryption", EncryptionEvent)
+
+            using EncryptionType = EncryptionEventContent::EncryptionType;
+
+            explicit EncryptionEvent(const QJsonObject& obj = {}) // TODO: apropriate default value
+                : StateEvent(typeId(), obj)
+            { }
+            EncryptionEvent(EncryptionEventContent&& c)
+                : StateEvent(typeId(), matrixTypeId(), c)
+            { }
+
+            /// A special constructor to create unknown EncryptionEvents
+            /**
+             * This is needed in order to use EncryptionEvent as a "base event
+             * class" in cases like GetEncryptionByRoomJob when EncryptionEvents
+             * (rather than RoomEvents or StateEvents) are resolved from JSON.
+             * For such cases loadEvent<> requires an underlying class to be
+             * constructible with unknownTypeId() instead of its genuine id.
+             * Don't use it directly.
+             * \sa GetEncryptionByRoomJob, loadEvent, unknownTypeId
+             */
+            EncryptionEvent(Type type, const QJsonObject& fullJson)
+                : StateEvent(type, fullJson)
+            { }
+
+            EncryptionType encryption() const  { return content().encryption; }
+
+            QString algorithm() const { return content().algorithm; }
+            int rotationPeriodMs() const { return content().rotationPeriodMs; }
+            int rotationPeriodMsgs() const { return content().rotationPeriodMsgs; }
+
+        private:
+            REGISTER_ENUM(EncryptionType)
+    };
+
+    template <>
+    class EventFactory<EncryptionEvent>
+    {
+        public:
+            static event_ptr_tt<EncryptionEvent> make(const QJsonObject& json,
+                                                      const QString&)
+            {
+                return makeEvent<EncryptionEvent>(json);
+            }
+    };
```

Comment:
This part is unnecessary for the same reason as above.

P.S. I probably should write a separate wiki page about `EventFactory<>`, given that my quick attempt to explain its idea here failed to the level I don't even include the text here.